File: packages/domain/src/public/advisory.ts
/**
 * Route advisory (ADR-0012): a sentence about a place and hours, from
 * expired reports, above thresholds, never a number and never a category.
 * Below the thresholds there is nothing — not *no data*, nothing — because
 * a sparse map of a city is a map of its poorer streets read as a claim
 * about the people who live there.
 */

export interface ExpiredReport {
  readonly account: string;
  readonly x: number;
  readonly y: number;
  readonly atMinutes: number;
}

/** One kilometre: coarse enough that a cell is an area, not an address. */
export const CELL_M = 1000;
export const WINDOW_MINUTES = 90 * 24 * 60;
export const MIN_REPORTS = 6;
export const MIN_ACCOUNTS = 4;
/** The band of hours that speaks: the shortest contiguous band holding at least this share of the cell's reports. */
export const BAND_SHARE = 0.6;

export interface Advisory {
  readonly cellX: number;
  readonly cellY: number;
  /** Hours of the day, inclusive start, exclusive end; may wrap midnight (from 21 to 2). */
  readonly fromHour: number;
  readonly toHour: number;
}

export function cellOf(x: number, y: number): { readonly cellX: number; readonly cellY: number } {
  return { cellX: Math.floor(x / CELL_M), cellY: Math.floor(y / CELL_M) };
}

function hourOf(atMinutes: number): number {
  return Math.floor((((atMinutes % 1440) + 1440) % 1440) / 60);
}

/**
 * The shortest band of hours, wrapping midnight if it must, that holds at
 * least `BAND_SHARE` of the counts. Ties go to the earlier start, so the
 * answer is one answer.
 */
export function band(byHour: ReadonlyArray<number>): { readonly fromHour: number; readonly toHour: number } {
  const total = byHour.reduce((a, b) => a + b, 0);
  let best: { fromHour: number; toHour: number; length: number } | null = null;
  for (let length = 1; length <= 24; length++) {
    for (let start = 0; start < 24; start++) {
      let sum = 0;
      for (let i = 0; i < length; i++) sum += byHour[(start + i) % 24] ?? 0;
      if (sum >= total * BAND_SHARE && (best === null || length < best.length)) {
        best = { fromHour: start, toHour: (start + length) % 24, length };
      }
    }
    if (best !== null) break;
  }
  return best ?? { fromHour: 0, toHour: 0 };
}

/**
 * What can be said about one cell, or null. Only reports inside the window
 * count; only a cell with enough reports from enough different accounts
 * speaks; and it says hours, not a count.
 */
export function advisoryFor(
  reports: ReadonlyArray<ExpiredReport>,
  cellX: number,
  cellY: number,
  nowMinutes: number,
): Advisory | null {
  const inCell = reports.filter((r) => {
    const c = cellOf(r.x, r.y);
    return c.cellX === cellX && c.cellY === cellY && nowMinutes - r.atMinutes <= WINDOW_MINUTES && r.atMinutes <= nowMinutes;
  });
  if (inCell.length < MIN_REPORTS) return null;
  if (new Set(inCell.map((r) => r.account)).size < MIN_ACCOUNTS) return null;
  const byHour = new Array<number>(24).fill(0);
  for (const r of inCell) byHour[hourOf(r.atMinutes)] = (byHour[hourOf(r.atMinutes)] ?? 0) + 1;
  const b = band(byHour);
  return { cellX, cellY, fromHour: b.fromHour, toHour: b.toHour };
}

/** The advisory for where the phone is, or null. */
export function advisoryAt(reports: ReadonlyArray<ExpiredReport>, x: number, y: number, nowMinutes: number): Advisory | null {
  const c = cellOf(x, y);
  return advisoryFor(reports, c.cellX, c.cellY, nowMinutes);
}

File: server/src/Sentinel.Domain/Advisory.cs
namespace Sentinel.Domain;

/// <summary>
/// Route advisory (ADR-0012), the C# mirror of <c>advisory.ts</c>: a place
/// and hours from expired reports, above thresholds, never a count and
/// never a category. Held to the TypeScript by <c>fixtures/advisory.json</c>.
/// </summary>
public static class Advisory
{
    public const int CellM = 1000;
    public const long WindowMinutes = 90L * 24 * 60;
    public const int MinReports = 6;
    public const int MinAccounts = 4;
    public const double BandShare = 0.6;

    public sealed record ExpiredReport(string Account, double X, double Y, long AtMinutes);
    public sealed record Result(int CellX, int CellY, int FromHour, int ToHour);

    public static (int CellX, int CellY) CellOf(double x, double y) => ((int)Math.Floor(x / CellM), (int)Math.Floor(y / CellM));

    private static int HourOf(long atMinutes) => (int)(((atMinutes % 1440) + 1440) % 1440 / 60);

    /// <summary>The shortest band holding at least the share, wrapping midnight if it must; ties to the earlier start.</summary>
    public static (int FromHour, int ToHour) Band(IReadOnlyList<int> byHour)
    {
        var total = byHour.Sum();
        for (var length = 1; length <= 24; length++)
        {
            (int from, int to)? best = null;
            for (var start = 0; start < 24; start++)
            {
                var sum = 0;
                for (var i = 0; i < length; i++) sum += byHour[(start + i) % 24];
                if (sum >= total * BandShare && best is null) best = (start, (start + length) % 24);
            }
            if (best is not null) return best.Value;
        }
        return (0, 0);
    }

    public static Result? For(IEnumerable<ExpiredReport> reports, int cellX, int cellY, long nowMinutes)
    {
        var inCell = reports.Where(r =>
        {
            var c = CellOf(r.X, r.Y);
            return c.CellX == cellX && c.CellY == cellY && nowMinutes - r.AtMinutes <= WindowMinutes && r.AtMinutes <= nowMinutes;
        }).ToList();
        if (inCell.Count < MinReports) return null;
        if (inCell.Select(r => r.Account).Distinct().Count() < MinAccounts) return null;
        var byHour = new int[24];
        foreach (var r in inCell) byHour[HourOf(r.AtMinutes)]++;
        var (from, to) = Band(byHour);
        return new Result(cellX, cellY, from, to);
    }

    public static Result? At(IEnumerable<ExpiredReport> reports, double x, double y, long nowMinutes)
    {
        var (cx, cy) = CellOf(x, y);
        return For(reports, cx, cy, nowMinutes);
    }
}
